### PR TITLE
Execute `versionCatalogUpdate` Gradle task from a CircleCI cron job and open PRs with version updates

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -6,6 +6,43 @@ commands:
...
+      - run:
+          name: Create a pull request if project files changed
+          # language=sh
+          command: |
+            if git diff-index --quiet HEAD; then
+              echo "No changes to project files after running the task, aborting"
+              exit 0
+            fi
+
+            branch_name="<<parameters.branch_name_prefix>>/$(<<parameters.branch_name_command>>)"
```
Comment (PawelLipski) [inline .circleci/config.yml:32]: `<<...>>`s are resolved by CircleCI before the script is executed

File: .circleci/config.yml [hunk 2]
```diff
@@ -6,6 +6,43 @@ commands:
...
+        type: string
+      commit_message_command:
+        type: string
+        default: 'echo ""'
+      labels:  # comma-separated
+        type: string
+    steps:
+      - run:
+          name: Create a pull request if project files changed
+          # language=sh
```
Comment (PawelLipski) [inline .circleci/config.yml:25]: For proper syntax highlight in IntelliJ (a feature called _language injection_)

File: build.gradle.kts [hunk 3]
```diff
@@ -39,16 +40,6 @@ val shouldRunAllCheckers by extra(isCI || project.hasProperty("runAllCheckers"))
 
 tasks.register<UpdateEapBuildNumber>("updateEapBuildNumber")
 
-tasks.withType<DependencyUpdatesTask> {
```
Comment (PawelLipski) [inline build.gradle.kts:42]: No longer needed, I've switched from `com.github.ben-manes.versions` Gradle plugin to `se.ascp.gradle.gradle-versions-filter`, which is almost identical but sets up this piece as a ready default

File: .circleci/config.yml [hunk 4]
```diff
@@ -178,35 +215,39 @@ jobs:
...
-              --no-edit \
-              --base=develop \
-              --labels=intellij-support
-
+            ./gradlew versionCatalogUpdate
+            cd buildSrc/
+            ../gradlew versionCatalogUpdate
+      - create_pull_request_if_files_changed:
+          branch_name_prefix: update-deps
+          branch_name_command: md5sum gradle/libs.versions.toml | cut -d' ' -f 1
```
Comment (PawelLipski) [inline .circleci/config.yml:235]: To have a unique branch name for each combination of versions

File: .circleci/config.yml [hunk 5]
```diff
@@ -178,35 +215,39 @@ jobs:
...
       - reconfigure_origin_remote
       - run:
-          name: Check if new eap build number has appeared
+          name: Check for updates of IntelliJ EAP
+          command: ./gradlew updateEapBuildNumber
+      - create_pull_request_if_files_changed:
+          branch_name_prefix: update-eap
+          branch_name_command: &extract_eap grep -Po '(?<=eapOfLatestSupportedMajor=).*(?=-EAP-SNAPSHOT)' intellijVersions.properties
+          commit_message_prefix: Update EAP build number to
+          commit_message_command: *extract_eap
```
Comment (PawelLipski) [inline .circleci/config.yml:224]: [An YAML anchor](https://www.linode.com/docs/guides/yaml-anchors-aliases-overrides-extensions/)

File: buildSrc/build.gradle.kts [hunk 6]
```diff
@@ -27,12 +29,16 @@ repositories {
...
-  implementation(libs.checkerFramework)
-  implementation(libs.jetbrains.intellij)
-  implementation(libs.jetbrains.kotlin)
   implementation(libs.jsoup)
-  implementation(libs.spotless)
+  // Needed so that versionCatalogUpdate, when executed against buildSrc,
+  // doesn't see jetbrains-annotations as an exceeded dependency and doesn't try to downgrade.
+  // See https://github.com/littlerobots/version-catalog-update-plugin#exceeded-dependencies
+  implementation(libs.jetbrains.annotations)
+  implementation(libs.pluginPackages.aspectj.postCompileWeaving)
```
Comment (PawelLipski) [inline buildSrc/build.gradle.kts:37]: Added `pluginPackages` segment for clarity

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt [hunk 7]
```diff
@@ -5,7 +5,9 @@ object IntellijVersions {
 
     private val intellijVersionsProp = IntellijVersionHelper.getProperties()
 
-    // When this value is updated, remember to update the minimum required IDEA version in README.md.
+    // When this value is updated, remember to update:
+    // 1. the minimum required IDEA version in README.md,
+    // 2. version of Gradle Kotlin plugin in gradle/libs.versions.toml
```
Comment (PawelLipski) [inline buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/IntellijVersions.kt:10]: That was always needed, just undocumented

File: buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt [hunk 8]
```diff
@@ -12,12 +12,6 @@ open class UpdateEapBuildNumber : DefaultTask() {
   @Internal
   val intellijSnapshotsUrl: String = "https://www.jetbrains.com/intellij-repository/snapshots/"
 
-  @Internal
-  @Option(
-      option = "fail-if-not-found",
```
Comment (PawelLipski) [inline buildSrc/src/main/kotlin/com/virtuslab/gitmachete/buildsrc/UpdateEapBuildNumber.kt:17]: No longer needed, I'm doing just `git diff-index --quiet HEAD` instead as it's more universal

File: gradle/libs.versions.toml [hunk 9]
```diff
@@ -62,4 +62,4 @@ jetbrains-kotlin = { id = "org.jetbrains.kotlin.jvm", version.ref = "plugins-jet
 spotless = { id = "com.diffplug.spotless", version.ref = "plugins-spotless" }
 taskTree = "com.dorongold.task-tree:2.1.0"
 versionCatalogUpdate = "nl.littlerobots.version-catalog-update:0.5.3"
-versions = "com.github.ben-manes.versions:0.42.0"
+versionsFilter = "se.ascp.gradle.gradle-versions-filter:0.1.16"
```
Comment (PawelLipski) [inline gradle/libs.versions.toml:65]: Same as `versions`, but with sane defaults

**Review by MaciejG604 — APPROVED** [review]

**Pull request merged**